In [1]:
import os
import sys
sys.path.append(os.path.abspath('..'))
from my.net import *
from task1_mnist import train_valid_data, run
data = train_valid_data()

#### Relu不同层数
| 模型架构 | 激活函数 |   损失函数   | 批大小 | 初始学习率 | 准确率97%迭代次数 | 最佳迭代次数 | 最佳准确率 |
| :------: | :------: | :----------: | :----: | :--------: | :---------------: | :----------: | :--------: |
|   两层   |   Relu   |     MSE      |  512   |     1      |                   |     120      |   96.89%   |
|   两层   |   Relu   | CrossEntropy |  512   |     1      |        28         |     109      |   97.38%   |
|   三层   |   Relu   | CrossEntropy |  512   |     1      |        81         |     137      |   97.17%   |
|   四层   |   Relu   | CrossEntropy |  512   |     1      |        114        |     138      |   97.06%   |

In [2]:
# 两层+Relu+CrossEntropy 512
net = Net(
    [
    {"module": Flatten},
    {"module": Mlp, "param": {"in_features": 784, "out_features": 128}},
    {"module": Relu},
    {"module": Mlp, "param": {"in_features": 128, "out_features": 10}},
    {"module": Softmax},
    ]
)
run(net,CrossEntropy, 1, 512, data)

 19%|█▊        | 149/800 [01:06<04:51,  2.23it/s, accuracy=97.36%, best_accuracy=97.38%, lr=0.0156, train_loss=0.00476, valid_loss=0.106]

模型连续40次无提升，提前终止训练
准确率97%所需迭代次数: 28
最佳迭代次数: 109
最佳准确率：97.38%
损失函数=CrossEntropy 学习率=0.015625 批大小=512
Net(
  (0): Flatten()
  (1): Mlp(in_features=784, out_features=128)
  (2): Relu()
  (3): Mlp(in_features=128, out_features=10)
  (4): Softmax()
)


In [3]:
# 三层+Relu+CrossEntropy 512
net = Net(
    [
    {"module": Flatten},
    {"module": Mlp, "param": {"in_features": 784, "out_features": 128}},
    {"module": Relu},
    {"module": Mlp, "param": {"in_features": 128, "out_features": 16}},
    {"module": Relu},
    {"module": Mlp, "param": {"in_features": 16, "out_features": 10}},
    {"module": Softmax},
    ]
)
run(net,CrossEntropy, 1, 512, data)

 22%|██▏       | 177/800 [01:12<04:15,  2.44it/s, accuracy=97.12%, best_accuracy=97.17%, lr=0.0156, train_loss=0.0179, valid_loss=0.106]

模型连续40次无提升，提前终止训练
准确率97%所需迭代次数: 81
最佳迭代次数: 137
最佳准确率：97.17%
损失函数=CrossEntropy 学习率=0.015625 批大小=512
Net(
  (0): Flatten()
  (1): Mlp(in_features=784, out_features=128)
  (2): Relu()
  (3): Mlp(in_features=128, out_features=16)
  (4): Relu()
  (5): Mlp(in_features=16, out_features=10)
  (6): Softmax()
)


In [4]:
# 四层+Relu+CrossEntropy 512
net = Net(
    [
    {"module": Flatten},
    {"module": Mlp, "param": {"in_features": 784, "out_features": 256}},
    {"module": Relu},
    {"module": Mlp, "param": {"in_features": 256, "out_features": 64}},
    {"module": Relu},
    {"module": Mlp, "param": {"in_features": 64, "out_features": 16}},
    {"module": Relu},
    {"module": Mlp, "param": {"in_features": 16, "out_features": 10}},
    {"module": Softmax},
    ]
)
run(net,CrossEntropy, 1, 512, data)

 22%|██▏       | 178/800 [02:14<07:49,  1.32it/s, accuracy=97.02%, best_accuracy=97.06%, lr=0.0625, train_loss=0.024, valid_loss=0.114] 

模型连续40次无提升，提前终止训练
准确率97%所需迭代次数: 114
最佳迭代次数: 138
最佳准确率：97.06%
损失函数=CrossEntropy 学习率=0.0625 批大小=512
Net(
  (0): Flatten()
  (1): Mlp(in_features=784, out_features=256)
  (2): Relu()
  (3): Mlp(in_features=256, out_features=64)
  (4): Relu()
  (5): Mlp(in_features=64, out_features=16)
  (6): Relu()
  (7): Mlp(in_features=16, out_features=10)
  (8): Softmax()
)


#### Gelu不同层数
| 模型架构 | 激活函数 |   损失函数   | 批大小 | 初始学习率 | 准确率97%迭代次数 | 最佳迭代次数 | 最佳准确率 |
| :------: | :------: | :----------: | :----: | :--------: | :---------------: | :----------: | :--------: |
|   两层   |   Gelu   | CrossEntropy |  512   |     1      |        27         |      76      |   97.35%   |
|   三层   |   Gelu   | CrossEntropy |  512   |     1      |        81         |     131      |   97.29%   |
|   四层   |   Gelu   | CrossEntropy |  512   |     1      |        115        |     122      |   97.04%   |

In [5]:
# 两层+Gelu+CrossEntropy 512
net = Net(
    [
    {"module": Flatten},
    {"module": Mlp, "param": {"in_features": 784, "out_features": 128}},
    {"module": Gelu},
    {"module": Mlp, "param": {"in_features": 128, "out_features": 10}},
    {"module": Softmax},
    ]
)
run(net,CrossEntropy, 1, 512, data)

 14%|█▍        | 116/800 [02:16<13:27,  1.18s/it, accuracy=97.32%, best_accuracy=97.35%, lr=0.0625, train_loss=0.00503, valid_loss=0.106]

模型连续40次无提升，提前终止训练
准确率97%所需迭代次数: 27
最佳迭代次数: 76
最佳准确率：97.35%
损失函数=CrossEntropy 学习率=0.0625 批大小=512
Net(
  (0): Flatten()
  (1): Mlp(in_features=784, out_features=128)
  (2): Gelu()
  (3): Mlp(in_features=128, out_features=10)
  (4): Softmax()
)


In [3]:
# 三层+Gelu+CrossEntropy 512
net = Net(
    [
    {"module": Flatten},
    {"module": Mlp, "param": {"in_features": 784, "out_features": 128}},
    {"module": Gelu},
    {"module": Mlp, "param": {"in_features": 128, "out_features": 16}},
    {"module": Gelu},
    {"module": Mlp, "param": {"in_features": 16, "out_features": 10}},
    {"module": Softmax},
    ]
)
run(net,CrossEntropy, 1, 512, data)

 21%|██▏       | 171/800 [03:45<13:48,  1.32s/it, accuracy=97.24%, best_accuracy=97.29%, lr=0.125, train_loss=0.00831, valid_loss=0.112]

模型连续40次无提升，提前终止训练
准确率97%所需迭代次数: 81
最佳迭代次数: 131
最佳准确率：97.29%
损失函数=CrossEntropy 学习率=0.125 批大小=512
Net(
  (0): Flatten()
  (1): Mlp(in_features=784, out_features=128)
  (2): Gelu()
  (3): Mlp(in_features=128, out_features=16)
  (4): Gelu()
  (5): Mlp(in_features=16, out_features=10)
  (6): Softmax()
)


In [4]:
# 四层+Gelu+CrossEntropy 512
net = Net(
    [
    {"module": Flatten},
    {"module": Mlp, "param": {"in_features": 784, "out_features": 256}},
    {"module": Gelu},
    {"module": Mlp, "param": {"in_features": 256, "out_features": 64}},
    {"module": Gelu},
    {"module": Mlp, "param": {"in_features": 64, "out_features": 16}},
    {"module": Gelu},
    {"module": Mlp, "param": {"in_features": 16, "out_features": 10}},
    {"module": Softmax},
    ]
)
run(net,CrossEntropy, 1, 512, data)

 20%|██        | 162/800 [08:13<32:23,  3.05s/it, accuracy=96.96%, best_accuracy=97.04%, lr=0.125, train_loss=0.0179, valid_loss=0.117]

模型连续40次无提升，提前终止训练
准确率97%所需迭代次数: 115
最佳迭代次数: 122
最佳准确率：97.04%
损失函数=CrossEntropy 学习率=0.125 批大小=512
Net(
  (0): Flatten()
  (1): Mlp(in_features=784, out_features=256)
  (2): Gelu()
  (3): Mlp(in_features=256, out_features=64)
  (4): Gelu()
  (5): Mlp(in_features=64, out_features=16)
  (6): Gelu()
  (7): Mlp(in_features=16, out_features=10)
  (8): Softmax()
)


| 模型架构 | 激活函数 |   损失函数   | 批大小 | 初始学习率 | 最佳迭代次数 | 最佳准确率 |
| :------: | :------: | :----------: | :----: | :--------: | :----------: | :--------: |
|   两层   | Sigmoid  | CrossEntropy |  512   |     20     |     109      |   96.38%   |
|   三层   | Sigmoid  | CrossEntropy |  512   |     20     |     126      |   96.13%   |

In [4]:
# 两层+Sigmoid+CrossEntropy 512
net = Net(
    [
    {"module": Flatten},
    {"module": Mlp, "param": {"in_features": 784, "out_features": 128}},
    {"module": Sigmoid},
    {"module": Mlp, "param": {"in_features": 128, "out_features": 10}},
    {"module": Softmax},
    ]
)
run(net,CrossEntropy, 20, 512, data)

 19%|█▊        | 149/800 [01:28<06:28,  1.68it/s, accuracy=96.17%, best_accuracy=96.38%, lr=2.5, train_loss=0.0209, valid_loss=0.124]

模型连续40次无提升，提前终止训练
准确率97%所需迭代次数: 0
最佳迭代次数: 109
最佳准确率：96.38%
损失函数=CrossEntropy 学习率=2.5 批大小=512
Net(
  (0): Flatten()
  (1): Mlp(in_features=784, out_features=128)
  (2): Sigmoid()
  (3): Mlp(in_features=128, out_features=10)
  (4): Softmax()
)


In [6]:
# 三层+Sigmoid+CrossEntropy 512
net = Net(
    [
    {"module": Flatten},
    {"module": Mlp, "param": {"in_features": 784, "out_features": 128}},
    {"module": Sigmoid},
    {"module": Mlp, "param": {"in_features": 128, "out_features": 16}},
    {"module": Sigmoid},
    {"module": Mlp, "param": {"in_features": 16, "out_features": 10}},
    {"module": Softmax},
    ]
)
run(net,CrossEntropy, 20, 512, data)

 21%|██        | 166/800 [01:44<06:39,  1.59it/s, accuracy=95.97%, best_accuracy=96.13%, lr=1.25, train_loss=0.0414, valid_loss=0.147]

模型连续40次无提升，提前终止训练
准确率97%所需迭代次数: 0
最佳迭代次数: 126
最佳准确率：96.13%
损失函数=CrossEntropy 学习率=1.25 批大小=512
Net(
  (0): Flatten()
  (1): Mlp(in_features=784, out_features=128)
  (2): Sigmoid()
  (3): Mlp(in_features=128, out_features=16)
  (4): Sigmoid()
  (5): Mlp(in_features=16, out_features=10)
  (6): Softmax()
)


| 模型架构 | 激活函数 |   损失函数   | 批大小 | 初始学习率 | 准确率97%迭代次数 | 最佳迭代次数 | 最佳准确率 |
| :------: | :------: | :----------: | :----: | :--------: | :---------------: | :----------: | :--------: |
|   两层   |   Relu   | CrossEntropy |  1024  |     1      |        107        |     168      |   97.15%   |
|   两层   |   Relu   | CrossEntropy |  512   |     1      |        28         |     109      |   97.38%   |
|   两层   |   Relu   | CrossEntropy |  256   |     1      |        14         |      46      |   97.45%   |
|   两层   |   Relu   | CrossEntropy |  128   |    0.1     |        16         |     106      |   97.51%   |
|   两层   |   Relu   | CrossEntropy |   64   |    0.1     |         5         |      33      |   97.83%   |

In [7]:
# 两层+Relu+MSE 512
net = Net(
    [
    {"module": Flatten},
    {"module": Mlp, "param": {"in_features": 784, "out_features": 128}},
    {"module": Relu},
    {"module": Mlp, "param": {"in_features": 128, "out_features": 10}},
    {"module": Softmax},
    ]
)
run(net,Mse, 1, 512, data)

 20%|██        | 160/800 [01:09<04:36,  2.31it/s, accuracy=96.89%, best_accuracy=96.89%, lr=0.00781, train_loss=0.00131, valid_loss=0.00488]

模型连续40次无提升，提前终止训练
准确率97%所需迭代次数: 0
最佳迭代次数: 120
最佳准确率：96.89%
损失函数=Mse 学习率=0.0078125 批大小=512
Net(
  (0): Flatten()
  (1): Mlp(in_features=784, out_features=128)
  (2): Relu()
  (3): Mlp(in_features=128, out_features=10)
  (4): Softmax()
)


In [8]:
# 两层+Relu+CrossEntropy 1024
net = Net(
    [
    {"module": Flatten},
    {"module": Mlp, "param": {"in_features": 784, "out_features": 128}},
    {"module": Relu},
    {"module": Mlp, "param": {"in_features": 128, "out_features": 10}},
    {"module": Softmax},
    ]
)
run(net,CrossEntropy, 1, 1024, data)

 26%|██▌       | 208/800 [01:17<03:40,  2.68it/s, accuracy=97.12%, best_accuracy=97.15%, lr=0.0312, train_loss=0.0211, valid_loss=0.0992]

模型连续40次无提升，提前终止训练
准确率97%所需迭代次数: 107
最佳迭代次数: 168
最佳准确率：97.15%
损失函数=CrossEntropy 学习率=0.03125 批大小=1024
Net(
  (0): Flatten()
  (1): Mlp(in_features=784, out_features=128)
  (2): Relu()
  (3): Mlp(in_features=128, out_features=10)
  (4): Softmax()
)


In [9]:
# 两层+Relu+CrossEntropy 256
net = Net(
    [
    {"module": Flatten},
    {"module": Mlp, "param": {"in_features": 784, "out_features": 128}},
    {"module": Relu},
    {"module": Mlp, "param": {"in_features": 128, "out_features": 10}},
    {"module": Softmax},
    ]
)
run(net,CrossEntropy, 1, 256, data)

 11%|█         | 86/800 [00:39<05:28,  2.18it/s, accuracy=97.42%, best_accuracy=97.45%, lr=0.125, train_loss=0.000863, valid_loss=0.131]

模型连续40次无提升，提前终止训练
准确率97%所需迭代次数: 14
最佳迭代次数: 46
最佳准确率：97.45%
损失函数=CrossEntropy 学习率=0.125 批大小=256
Net(
  (0): Flatten()
  (1): Mlp(in_features=784, out_features=128)
  (2): Relu()
  (3): Mlp(in_features=128, out_features=10)
  (4): Softmax()
)


In [16]:
# 两层+Relu+CrossEntropy 128
net = Net(
    [
    {"module": Flatten},
    {"module": Mlp, "param": {"in_features": 784, "out_features": 128}},
    {"module": Relu},
    {"module": Mlp, "param": {"in_features": 128, "out_features": 10}},
    {"module": Softmax},
    ]
)
run(net,CrossEntropy, 0.1, 128, data)

 18%|█▊        | 146/800 [01:40<07:30,  1.45it/s, accuracy=97.49%, best_accuracy=97.51%, lr=0.00625, train_loss=0.00171, valid_loss=0.111]

模型连续40次无提升，提前终止训练
准确率97%所需迭代次数: 16
最佳迭代次数: 106
最佳准确率：97.51%
损失函数=CrossEntropy 学习率=0.00625 批大小=128
Net(
  (0): Flatten()
  (1): Mlp(in_features=784, out_features=128)
  (2): Relu()
  (3): Mlp(in_features=128, out_features=10)
  (4): Softmax()
)


In [14]:
# 两层+Relu+CrossEntropy 64
net = Net(
    [
    {"module": Flatten},
    {"module": Mlp, "param": {"in_features": 784, "out_features": 128}},
    {"module": Relu},
    {"module": Mlp, "param": {"in_features": 128, "out_features": 10}},
    {"module": Softmax},
    ]
)
run(net,CrossEntropy, 0.1, 64, data)

  9%|▉         | 73/800 [01:41<16:50,  1.39s/it, accuracy=97.78%, best_accuracy=97.83%, lr=0.0125, train_loss=0.000488, valid_loss=0.116]

模型连续40次无提升，提前终止训练
准确率97%所需迭代次数: 5
最佳迭代次数: 33
最佳准确率：97.83%
损失函数=CrossEntropy 学习率=0.0125 批大小=64
Net(
  (0): Flatten()
  (1): Mlp(in_features=784, out_features=128)
  (2): Relu()
  (3): Mlp(in_features=128, out_features=10)
  (4): Softmax()
)


| 模型架构 | 激活函数 |   损失函数   | 批大小 | 初始学习率 | 最佳迭代次数 | 最佳准确率 |
| :------: | :------: | :----------: | :----: | :--------: | :----------: | :--------: |
|   两层   |   Relu   | CrossEntropy |  512   |     1      |     109      |   97.38%   |
|   两层   |   Relu   | CrossEntropy |  512   |    0.1     |     279      |   96.82%   |
|   两层   |   Relu   | CrossEntropy |  512   |    0.05    |     296      |   96.53%   |

In [ ]:
# 两层+Relu+CrossEntropy 0.1
net = Net(
    [
    {"module": Flatten},
    {"module": Mlp, "param": {"in_features": 784, "out_features": 128}},
    {"module": Relu},
    {"module": Mlp, "param": {"in_features": 128, "out_features": 10}},
    {"module": Softmax},
    ]
)
run(net,CrossEntropy, 0.1, 512, data)

 40%|███▉      | 319/800 [02:03<03:05,  2.59it/s, accuracy=96.81%, best_accuracy=96.82%, lr=0.00156, train_loss=0.0423, valid_loss=0.106]

模型连续40次无提升，提前终止训练
准确率97%所需迭代次数: 0
最佳迭代次数: 279
最佳准确率：96.82%
损失函数=CrossEntropy 学习率=0.0015625 批大小=512
Net(
  (0): Flatten()
  (1): Mlp(in_features=784, out_features=128)
  (2): Relu()
  (3): Mlp(in_features=128, out_features=10)
  (4): Softmax()
)


In [ ]:
# 两层+Relu+CrossEntropy 0.05
net = Net(
    [
    {"module": Flatten},
    {"module": Mlp, "param": {"in_features": 784, "out_features": 128}},
    {"module": Relu},
    {"module": Mlp, "param": {"in_features": 128, "out_features": 10}},
    {"module": Softmax},
    ]
)
run(net,CrossEntropy, 0.05, 512, data)

 42%|████▏     | 336/800 [02:11<03:01,  2.55it/s, accuracy=96.47%, best_accuracy=96.53%, lr=0.00313, train_loss=0.0681, valid_loss=0.118]

模型连续40次无提升，提前终止训练
准确率97%所需迭代次数: 0
最佳迭代次数: 296
最佳准确率：96.53%
损失函数=CrossEntropy 学习率=0.003125 批大小=512
Net(
  (0): Flatten()
  (1): Mlp(in_features=784, out_features=128)
  (2): Relu()
  (3): Mlp(in_features=128, out_features=10)
  (4): Softmax()
)


| 模型架构 | 消融率 | 激活函数 |   损失函数   | 批大小 | 初始学习率 | 准确率97%迭代次数 | 最佳迭代次数 | 最佳准确率 |
| :------: | :----: | :------: | :----------: | :----: | :--------: | :---------------: | :----------: | :--------: |
|   三层   |   0    |   Relu   | CrossEntropy |  512   |     1      |        81         |     137      |   97.17%   |
|   三层   |  0.1   |   Relu   | CrossEntropy |  512   |     1      |        81         |     200      |   97.46%   |
|   三层   |  0.15  |   Relu   | CrossEntropy |  512   |     1      |        80         |     172      |   97.52%   |
|   三层   |  0.2   |   Relu   | CrossEntropy |  512   |     1      |        79         |     179      |   97.47%   |
|   三层   |  0.3   |   Relu   | CrossEntropy |  512   |     1      |        105        |     177      |   97.33%   |
|   三层   |  0.4   |   Relu   | CrossEntropy |  512   |     1      |                   |     147      |   96.83%   |


In [ ]:
# 三层+Relu+CrossEntropy Dropout=0.1
net = Net(
    [
    {"module": Flatten},
    {"module": Mlp, "param": {"in_features": 784, "out_features": 128}},
    {"module": Relu},
    {"module": Dropout, "param": {"p": 0.1}},
    {"module": Mlp, "param": {"in_features": 128, "out_features": 16}},
    {"module": Relu},
    {"module": Dropout, "param": {"p": 0.1}},
    {"module": Mlp, "param": {"in_features": 16, "out_features": 10}},
    {"module": Softmax},
    ]
)
run(net,CrossEntropy, 1, 512, data)

 30%|███       | 240/800 [01:47<04:09,  2.24it/s, accuracy=97.42%, best_accuracy=97.46%, lr=0.00391, train_loss=0.0182, valid_loss=0.0917]

模型连续40次无提升，提前终止训练
准确率97%所需迭代次数: 82
最佳迭代次数: 200
最佳准确率：97.46%
损失函数=CrossEntropy 学习率=0.00390625 批大小=512
Net(
  (0): Flatten()
  (1): Mlp(in_features=784, out_features=128)
  (2): Relu()
  (3): Dropout(p=0.1)
  (4): Mlp(in_features=128, out_features=16)
  (5): Relu()
  (6): Dropout(p=0.1)
  (7): Mlp(in_features=16, out_features=10)
  (8): Softmax()
)


In [2]:
# 三层+Relu+CrossEntropy Dropout=0.15
net = Net(
    [
    {"module": Flatten},
    {"module": Mlp, "param": {"in_features": 784, "out_features": 128}},
    {"module": Relu},
    {"module": Dropout, "param": {"p": 0.15}},
    {"module": Mlp, "param": {"in_features": 128, "out_features": 16}},
    {"module": Relu},
    {"module": Dropout, "param": {"p": 0.15}},
    {"module": Mlp, "param": {"in_features": 16, "out_features": 10}},
    {"module": Softmax},
    ]
)
run(net,CrossEntropy, 1, 512, data)

 26%|██▋       | 212/800 [02:31<06:59,  1.40it/s, accuracy=97.48%, best_accuracy=97.52%, lr=0.0312, train_loss=0.0177, valid_loss=0.093] 

模型连续40次无提升，提前终止训练
准确率97%所需迭代次数: 80
最佳迭代次数: 172
最佳准确率：97.52%
损失函数=CrossEntropy 学习率=0.03125 批大小=512
Net(
  (0): Flatten()
  (1): Mlp(in_features=784, out_features=128)
  (2): Relu()
  (3): Dropout(p=0.15)
  (4): Mlp(in_features=128, out_features=16)
  (5): Relu()
  (6): Dropout(p=0.15)
  (7): Mlp(in_features=16, out_features=10)
  (8): Softmax()
)


In [ ]:
# 三层+Relu+CrossEntropy Dropout=0.2
net = Net(
    [
    {"module": Flatten},
    {"module": Mlp, "param": {"in_features": 784, "out_features": 128}},
    {"module": Relu},
    {"module": Dropout, "param": {"p": 0.2}},
    {"module": Mlp, "param": {"in_features": 128, "out_features": 16}},
    {"module": Relu},
    {"module": Dropout, "param": {"p": 0.2}},
    {"module": Mlp, "param": {"in_features": 16, "out_features": 10}},
    {"module": Softmax},
    ]
)
run(net,CrossEntropy, 1, 512, data)

 27%|██▋       | 219/800 [02:05<05:34,  1.74it/s, accuracy=97.45%, best_accuracy=97.47%, lr=0.0156, train_loss=0.0217, valid_loss=0.0973]

模型连续40次无提升，提前终止训练
准确率97%所需迭代次数: 79
最佳迭代次数: 179
最佳准确率：97.47%
损失函数=CrossEntropy 学习率=0.015625 批大小=512
Net(
  (0): Flatten()
  (1): Mlp(in_features=784, out_features=128)
  (2): Relu()
  (3): Dropout(p=0.2)
  (4): Mlp(in_features=128, out_features=16)
  (5): Relu()
  (6): Dropout(p=0.2)
  (7): Mlp(in_features=16, out_features=10)
  (8): Softmax()
)


In [ ]:
# 三层+Relu+CrossEntropy Dropout=0.3
net = Net(
    [
    {"module": Flatten},
    {"module": Mlp, "param": {"in_features": 784, "out_features": 128}},
    {"module": Relu},
    {"module": Dropout, "param": {"p": 0.3}},
    {"module": Mlp, "param": {"in_features": 128, "out_features": 16}},
    {"module": Relu},
    {"module": Dropout, "param": {"p": 0.3}},
    {"module": Mlp, "param": {"in_features": 16, "out_features": 10}},
    {"module": Softmax},
    ]
)
run(net,CrossEntropy, 1, 512, data)

 27%|██▋       | 217/800 [01:37<04:22,  2.22it/s, accuracy=97.29%, best_accuracy=97.33%, lr=0.0625, train_loss=0.0285, valid_loss=0.109]

模型连续40次无提升，提前终止训练
准确率97%所需迭代次数: 105
最佳迭代次数: 177
最佳准确率：97.33%
损失函数=CrossEntropy 学习率=0.0625 批大小=512
Net(
  (0): Flatten()
  (1): Mlp(in_features=784, out_features=128)
  (2): Relu()
  (3): Dropout(p=0.3)
  (4): Mlp(in_features=128, out_features=16)
  (5): Relu()
  (6): Dropout(p=0.3)
  (7): Mlp(in_features=16, out_features=10)
  (8): Softmax()
)


In [ ]:
# 三层+Relu+CrossEntropy Dropout=0.4
net = Net(
    [
    {"module": Flatten},
    {"module": Mlp, "param": {"in_features": 784, "out_features": 128}},
    {"module": Relu},
    {"module": Dropout, "param": {"p": 0.4}},
    {"module": Mlp, "param": {"in_features": 128, "out_features": 16}},
    {"module": Relu},
    {"module": Dropout, "param": {"p": 0.4}},
    {"module": Mlp, "param": {"in_features": 16, "out_features": 10}},
    {"module": Softmax},
    ]
)
run(net,CrossEntropy, 1, 512, data)

 23%|██▎       | 187/800 [01:24<04:36,  2.22it/s, accuracy=96.73%, best_accuracy=96.83%, lr=0.0312, train_loss=0.0571, valid_loss=0.128]

模型连续40次无提升，提前终止训练
准确率97%所需迭代次数: 0
最佳迭代次数: 147
最佳准确率：96.83%
损失函数=CrossEntropy 学习率=0.03125 批大小=512
Net(
  (0): Flatten()
  (1): Mlp(in_features=784, out_features=128)
  (2): Relu()
  (3): Dropout(p=0.4)
  (4): Mlp(in_features=128, out_features=16)
  (5): Relu()
  (6): Dropout(p=0.4)
  (7): Mlp(in_features=16, out_features=10)
  (8): Softmax()
)
